plugins/inputs/tencentcloudcm/tencentcloudcm.go

```diff
@@ -0,0 +1,448 @@
+
+	// construct metric object
+	for i := range t.Accounts {
+		for j := range t.Accounts[i].Namespaces {
+			for k := range t.Accounts[i].Namespaces[j].Regions {
+				for l := range t.Accounts[i].Namespaces[j].Metrics {
+					instances := t.Accounts[i].Namespaces[j].Regions[k].Instances
+					if len(instances) == 0 {
+						instances = t.discoverTool.GetInstances(t.Accounts[i].Name, t.Accounts[i].Namespaces[j].Name, t.Accounts[i].Namespaces[j].Regions[k].RegionName)
+					}
+					if len(instances) == 0 {
+						continue
+					}
+					metricObjects = append(metricObjects, MetricObject{
+						t.Accounts[i].Namespaces[j].Metrics[l],
+						t.Accounts[i].Namespaces[j].Regions[k].RegionName,
+						t.Accounts[i].Namespaces[j].Name,
+						t.Accounts[i],
+						instances,
+					})
+				}
+			}
+		}
+	}
+	return metricObjects
+}
+
+func (t *TencentCloudCM) Gather(acc telegraf.Accumulator) error {
+
+	t.updateWindow(time.Now())
+
+	metricObjects := t.client.GetMetricObjects(*t)
+
+	lmtr := limiter.NewRateLimiter(t.RateLimit, time.Second)
+	defer lmtr.Stop()
+
+	wg := sync.WaitGroup{}
+	rLock := sync.Mutex{}
+	results := []GetMonitorDataResponse{}
+
+	// requestIDMap contains request ID and metric objects for later aggregation
+	requestIDMap := map[string]MetricObject{}
+
+	for i := range metricObjects {
+
+		wg.Add(1)
+		<-lmtr.C
+		go func(m MetricObject) {
+			defer wg.Done()
+
+			client := t.client.NewClient(m.Region, m.Account.Crs, *t)
+			request := t.client.NewGetMonitorDataRequest(m.Namespace, m.Metric, m.Instances, *t)
+
+			result, err := t.client.GatherMetrics(client, request, *t)
+			if err != nil {
+				acc.AddError(err)
+				return
+			}
+
+			rLock.Lock()
+			requestIDMap[*result.Response.RequestId] = m
+			rLock.Unlock()
+
+			rLock.Lock()
+			results = append(results, *result)
+			rLock.Unlock()
+
+		}(metricObjects[i])
+
+	}
+	wg.Wait()
+
+	for i := range results {
+		for j := range results[i].Response.DataPoints {
+			tags := map[string]string{}
+			for k := range results[i].Response.DataPoints[j].Dimensions {
+				tags[*results[i].Response.DataPoints[j].Dimensions[k].Name] = *results[i].Response.DataPoints[j].Dimensions[k].Value
+			}
+			metricObject := requestIDMap[*results[i].Response.RequestId]
+			tags["account"] = metricObject.Account.Name
+			tags["namespace"] = metricObject.Namespace
+			tags["region"] = metricObject.Region
+			tags["period"] = fmt.Sprintf("%v", *results[i].Response.Period)
+			tags["metric"] = *results[i].Response.MetricName
+			tags["request_id"] = *results[i].Response.RequestId
+
+			for index := range results[i].Response.DataPoints[j].Values {
+				acc.AddFields(metricObject.Namespace, map[string]interface{}{"value": results[i].Response.DataPoints[j].Values[index]}, tags, time.Unix(int64(*results[i].Response.DataPoints[j].Timestamps[index]), 0).Local())
+			}
+
+		}
+	}
+
+	return nil
+}
+
+func init() {
+	inputs.Add("tencentcloudcm", func() telegraf.Input {
+		return New()
```

Comment:
It's a matter of taste, but why not just return the struct pointer instead of wrapping it in a `New()` function?

Reply:
fixed for cleanness.